replica/table.cc

```diff
@@ -1349,30 +1349,30 @@ future<> table::snapshot(database& db, sstring name) {
     auto jsondir = _config.datadir + "/snapshots/" + name;
     tlogger.debug("snapshot {}", jsondir);
 
-    // FIXME: indentation
-       return with_semaphore(_sstable_deletion_sem, 1, [this, &db, jsondir = std::move(jsondir)]() {
+    auto sstable_deletion_guard = co_await get_units(_sstable_deletion_sem, 1);
+
+        // FIXME: indentation
         auto tables = boost::copy_range<std::vector<sstables::shared_sstable>>(*_sstables->all());
-        return do_with(std::move(tables), std::move(jsondir), [this, &db] (std::vector<sstables::shared_sstable>& tables, const sstring& jsondir) {
-            return io_check([&jsondir] { return recursive_touch_directory(jsondir); }).then([this, &db, &jsondir, &tables] {
-                return max_concurrent_for_each(tables, db.get_config().initial_sstable_loading_concurrency(), [&db, &jsondir] (sstables::shared_sstable sstable) {
+        co_await io_check([&jsondir] { return recursive_touch_directory(jsondir); });
+        co_await max_concurrent_for_each(tables, db.get_config().initial_sstable_loading_concurrency(), [&db, &jsondir] (sstables::shared_sstable sstable) {
                   return with_semaphore(db.get_sharded_sst_dir_semaphore().local(), 1, [&jsondir, sstable] {
                     return io_check([sstable, &dir = jsondir] {
                         return sstable->create_links(dir);
                     });
                   });
                 });
-            }).then([&jsondir, &tables] {
-                return io_check(sync_directory, jsondir);
-            }).finally([this, &tables, &db, &jsondir] {
+        co_await io_check(sync_directory, jsondir);
+
```

Comment:
I don't understand your comment about "catch immediate errors and return them". Here, if anything above here throws, the entire thing will throw, and the code below will not run. This is different from the code before your modification, which had a "finally" on the code below.

Reply:
Yes.  This is the gist of this whole series.
The existing code had a finally clause bu it didn't handle all errors.

Any error in this part: https://github.com/scylladb/scylla/blob/eeec502aee7a693ccaf84bbf4a4313350c1a26bd/replica/table.cc#L1366-L1383

would cause other shards to hang, since the semaphore wouldn't be signaled here:
https://github.com/scylladb/scylla/blob/eeec502aee7a693ccaf84bbf4a4313350c1a26bd/replica/table.cc#L1386

and here: https://github.com/scylladb/scylla/blob/eeec502aee7a693ccaf84bbf4a4313350c1a26bd/replica/table.cc#L1404

Reply:
What confused me was that you lost the "finally" stuff and re-added it only in the next patch - which I didn't know would happen until I moved to reviewing the next patch.

Reply:
I squashed them together in v2